### PR TITLE
fix(cli): Fix detection --app, --region, --remote and --addon from command line arguments

### PR DESCRIPTION
**Pull request by aurelien-reeves-scalingo**
Fix https://github.com/Scalingo/cli/issues/780

- [x] Update CHANGELOG.md

### CODE REVIEW

**Comment by aurelien-reeves-scalingo** [comment]
The issue may be due to default values

There may be other flags to take care of. On a first glance, the `addon` one may be impacted as it use a similar approach.

File: detect/app.go [hunk 1]
```diff
@@ -60,6 +63,16 @@ func GetAppNameFromGitRemote(directory string, remoteName string) (string, error
 	return "", errgo.Newf("[detect] Scalingo Git remote hasn't been found")
 }
 
+func extractAppNameFromCommandLine(c *cli.Context) string {
+	for _, context := range c.Lineage() {
```
Comment (EtienneM) [inline detect/app.go:67]: `context` is not a very good variable name as it is often the name of an imported package. Suggestion:

```suggestion
	for _, cliContext := range c.Lineage() {
```

Comment (EtienneM) [inline detect/app.go:67]: I'm not sure I understand what does `Lineage` do (https://pkg.go.dev/github.com/urfave/cli/v2#Context.Lineage), but I trust you if you tested it and it fixes the problem ^^

Shouldn't we also use it for the other global flags such as the region? I think it has the same problem, don't you think?

**Review by EtienneM — CHANGES_REQUESTED** [review]
nitpick:
- missing changelog
- PR is in draft

File: detect/app.go (hunk 1 above)
Comment (aurelien-reeves-scalingo) [reply detect/app.go]: There are two "app" flags defined: one global option, and one local to the actual command, both with default value "<name>".

`c.String("name")` in the context of a command returns the value for the local flag. When it is not set in the command line, it is the default value.

`c.Lineage()` returns an array with the local context, and all the parent contexts before, thus the root one with the global options.

So we can loop through that array to look for an `app` flag which has actually been set

Regarding the other `global` flags, indeed we have to investigate. I think the `addon` may be imacted as it uses a similar approach as the `app` one. For the other ones, more investigation are required

**Comment by aurelien-reeves-scalingo** [comment]
Same issue with `--addon`. Working on it

**Comment by aurelien-reeves-scalingo** [comment]
`--addon`: done

The `--remote` flag should be impacted too. However that flag is used only with `create` and `git-setup`. What about removing it from the global options? @EtienneM

**Comment by aurelien-reeves-scalingo** [comment]
Changelog has been updated not with the addition of an entry, but with the removal of a previous breaking change notice

**Comment by aurelien-reeves-scalingo** [comment]
> The `--remote` flag should be impacted too. However that flag is used only with `create` and `git-setup`, and no documentation refers to it as a global option. What about removing it from the global options? @EtienneM

After discussion with @Soulou, it is actually used with the app detection based on the remote.

```console
# The following command will detect the application name based on the given remote
$ scalingo --remote scalingo ps
```

So, keep the support for it, and still to be fixed with urfave/cli v2

File: detect/app.go [hunk 2]
```diff
@@ -63,10 +63,21 @@ func GetAppNameFromGitRemote(directory string, remoteName string) (string, error
 	return "", errgo.Newf("[detect] Scalingo Git remote hasn't been found")
 }
 
+// Returns the remote name specified in command flags
```
Comment (EtienneM) [inline detect/app.go:66]: Just a picky comment: in Go, function comments start with the function name

```suggestion
// RemoteNameFromFlags returns the remote name specified in command flags
```

This is for a prettier output in the Go doc. For example: https://pkg.go.dev/github.com/Scalingo/go-scalingo/v5#NewStaticTokenGenerator

**Review by EtienneM — APPROVED** [review]
Just a remark but I already validate it :)

**Pull request merged**